### PR TITLE
[MAR-3119]

### PR DESCRIPTION
**Pull request by resh-ie**

### CODE REVIEW

**Comment by malmas18** [comment]
remove all `.DS_Store` files add it to the `.gitignore` file

File: .manifests/logger.alpha.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+applications:
+- name: dm-alpha-logger
+  buildpack: dotnet_core_buildpack
+  path: ../subscribers/logs/logger
+  health-check-type: process
+  instances: 1
+  memory: 128M
```
Comment (malmas18) [inline .manifests/logger.alpha.yml]:
```suggestion
  memory: 50M
```

File: subscribers/logs/logger.tests/LoggerTests.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    public class LoggerTests {
+
+        [Fact]
+        public void Can_log_information_message() {
+            // Arrange
+            var logger = new Mock<ILoggerAdapter<AppService>>();
+            var dbSet = new Mock<DbSet<LogEntry>>();
+            var connection = new SqliteConnection("DataSource=:memory:");
+            connection.Open();
```
Comment (malmas18) [inline subscribers/logs/logger.tests/LoggerTests.cs:25]: need to close the connection also. checkout https://docs.microsoft.com/en-us/ef/core/miscellaneous/testing/sqlite
`
try {
CODE
} finally {
connection.close()
}
`

File: .manifests/logger.prod.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+applications:
+- name: dm-logger
+  buildpack: dotnet_core_buildpack
+  path: ../subscribers/logs/logger
+  health-check-type: process
+  instances: 1
+  memory: 50M
```
Comment (malmas18) [inline .manifests/logger.prod.yml]:
```suggestion
  memory: 128M
```

File: subscribers/logs/logger/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Slack Notifier
```
Comment (malmas18) [inline subscribers/logs/logger/README.md]:
```suggestion
# SNS logger
```

File: subscribers/slack/worker.tests/ApplicationMessageProcessorTest.cs [hunk 5]
```diff
@@ -11,6 +11,7 @@
 using Dta.Marketplace.Subscribers.Slack.Worker.Services;
 
 namespace Dta.Marketplace.Subscribers.Slack.Worker.Tests {
+
```
Comment (malmas18) [inline subscribers/slack/worker.tests/ApplicationMessageProcessorTest.cs]: unrelated change
```suggestion
```

File: subscribers/slack/worker/AppService.cs [hunk 6]
```diff
@@ -69,6 +69,7 @@ public class AppService : IHostedService, IDisposable {
             var receiveMessageResponse = await _sqsClient.ReceiveMessageAsync(receiveMessageRequest);
             foreach (var message in receiveMessageResponse.Messages) {
                 _logger.LogDebug("Message Id: {MessageId}", message.MessageId);
+                _logger.LogInformation(message.Body);
```
Comment (malmas18) [inline subscribers/slack/worker/AppService.cs]:
```suggestion
```

File: subscribers/slack/worker.tests/UserMessageProcessorTest.cs [hunk 7]
```diff
@@ -1,9 +1,9 @@
+using System;
```
Comment (malmas18) [inline subscribers/slack/worker.tests/UserMessageProcessorTest.cs]: unrelated changes in this file

File: subscribers/slack/worker/worker.csproj [hunk 8]
```diff
@@ -27,5 +27,6 @@
     <PackageReference Include="Serilog" Version="2.8.0" />
     <PackageReference Include="Serilog.Extensions.Hosting" Version="2.0.0" />
     <PackageReference Include="Serilog.Sinks.Console" Version="3.1.1" />
+    <PackageReference Include="System" Version="4.1.0311.2" />
```
Comment (malmas18) [inline subscribers/slack/worker/worker.csproj]:
```suggestion
```

Comment (resh-ie) [reply subscribers/slack/worker/worker.csproj]: done

File: subscribers/slack/worker.tests/UserMessageProcessorTest.cs (hunk 7 above)
Comment (resh-ie) [reply subscribers/slack/worker.tests/UserMessageProcessorTest.cs]: done

File: subscribers/slack/worker/AppService.cs (hunk 6 above)
Comment (resh-ie) [reply subscribers/slack/worker/AppService.cs]: done

File: subscribers/logs/logger/README.md (hunk 4 above)
Comment (resh-ie) [reply subscribers/logs/logger/README.md]: done

File: .manifests/logger.prod.yml (hunk 3 above)
Comment (resh-ie) [reply .manifests/logger.prod.yml]: done

File: subscribers/logs/logger.tests/LoggerTests.cs (hunk 2 above)
Comment (resh-ie) [reply subscribers/logs/logger.tests/LoggerTests.cs:25]: done

File: .manifests/logger.alpha.yml (hunk 1 above)
Comment (resh-ie) [reply .manifests/logger.alpha.yml]: done

File: subscribers/slack/worker.tests/ApplicationMessageProcessorTest.cs (hunk 5 above)
Comment (resh-ie) [reply subscribers/slack/worker.tests/ApplicationMessageProcessorTest.cs]: > .DS_Store

done

File: subscribers/logs/logger.tests/LoggerTests.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+using Amazon.SQS.Model;
+
+using Microsoft.EntityFrameworkCore;
+
+namespace Dta.Marketplace.Subscribers.Logger.Worker.Logger.Tests {
+
+    public class LoggerTests {
+
+        [Fact]
+        public void Can_log_information_message() {
```
Comment (tobiasbrown) [inline subscribers/logs/logger.tests/LoggerTests.cs]: The naming of methods / functions should be Pascal case.  Check [Microsoft's conventions](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions).
```suggestion
        public void CanLogInformationMessage() {
```

File: subscribers/logs/logger.tests/LoggerTests.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+            // Arrange
+            var logger = new Mock<ILoggerAdapter<AppService>>();
+            var dbSet = new Mock<DbSet<LogEntry>>();
+            var connection = new SqliteConnection("DataSource=:memory:");
+            connection.Open();
+            try {
+            //In-memory database only exists while the connection is open
```
Comment (tobiasbrown) [inline subscribers/logs/logger.tests/LoggerTests.cs]: Looks like the code wrapped by the `try` needs to be indented.

File: subscribers/logs/logger.tests/LoggerTests.cs [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+                    Body = json
+                });
+            }
+
+            // Assert
+            using (var context = new LoggerContext(options)) {
+                logger.Verify(l => l.LogInformation(json), Times.Once);
+            }
+        } 
+        finally{
```
Comment (tobiasbrown) [inline subscribers/logs/logger.tests/LoggerTests.cs]:
```suggestion
        finally {
```

File: subscribers/logs/logger.tests/logger.tests.csproj [hunk 12]
```diff
@@ -0,0 +1,24 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+
```
Comment (tobiasbrown) [inline subscribers/logs/logger.tests/logger.tests.csproj]:
```suggestion
```

File: subscribers/logs/logger/AppService.cs [hunk 13]
```diff
@@ -0,0 +1,87 @@
+/* */
```
Comment (tobiasbrown) [inline subscribers/logs/logger/AppService.cs]:
```suggestion
```

File: subscribers/logs/logger/AppService.cs [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                _config.Value.AwsSqsRegion,
+                _config.Value.AwsSqsServiceUrl,
+                _config.Value.AwsSqsQueueUrl,
+                sentryEnabled = string.IsNullOrWhiteSpace(_config.Value.SentryDsn) ? false : true
+            });
+
+            var sqsConfig = new AmazonSQSConfig {
+                RegionEndpoint = RegionEndpoint.GetBySystemName(_config.Value.AwsSqsRegion)
+            };
+            if (string.IsNullOrWhiteSpace(_config.Value.AwsSqsServiceUrl) == false) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/AppService.cs:42]: It might be better to use `!` to test for `false` values.  I can see you've used it elsewhere in this PR.
```suggestion
            if (!string.IsNullOrWhiteSpace(_config.Value.AwsSqsServiceUrl)) {
```

File: subscribers/logs/logger/AppService.cs [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+                string.IsNullOrWhiteSpace(_config.Value.AwsSqsAccessKeyId) == false) {
+                _sqsClient = new AmazonSQSClient(_config.Value.AwsSqsAccessKeyId, _config.Value.AwsSqsSecretAccessKey, sqsConfig);
+            } else {
+                _sqsClient = new AmazonSQSClient(sqsConfig);
+            }
+
+            _timer = new Timer(DoWork, null, TimeSpan.Zero, TimeSpan.FromSeconds(_config.Value.WorkIntervalInSeconds));
+            return Task.CompletedTask;
+        }
+        public Task StopAsync(CancellationToken cancellationToken) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/AppService.cs:56]:
```suggestion
        
        public Task StopAsync(CancellationToken cancellationToken) {
```

File: subscribers/logs/logger/AppService.cs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            _logger.LogInformation("Stopping daemon.");
+            _timer?.Change(Timeout.Infinite, 0);
+            return Task.CompletedTask;
+        }
+        public void Dispose() {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/AppService.cs:62]:
```suggestion

        public void Dispose() {
```

File: subscribers/logs/logger/AppService.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+            _logger.LogInformation("Disposing....");
+            _timer?.Dispose();
+            _sqsClient?.Dispose();
+        }
+        private async void DoWork(object state) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/AppService.cs:68]:
```suggestion

        private async void DoWork(object state) {
```

File: subscribers/logs/logger/AppService.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+            var receiveMessageRequest = new ReceiveMessageRequest() {
+                QueueUrl = _config.Value.AwsSqsQueueUrl,
+                WaitTimeSeconds = _config.Value.AwsSqsLongPollTimeInSeconds
+            };
+            _logger.LogDebug("Heartbeat: {Now}", DateTime.Now);
+            var receiveMessageResponse = await _sqsClient.ReceiveMessageAsync(receiveMessageRequest);
+                foreach (var message in receiveMessageResponse.Messages) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/AppService.cs]: Indentation of this for loop looks off.  Might need to come back.

File: subscribers/logs/logger/MessageProcessor.cs [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+using Amazon.SQS.Model;
+
+namespace Dta.Marketplace.Subscribers.Logger.Worker
+{
+    public class MessageProcessor: Exception, IMessageProcessor
+    {
+        private readonly ILoggerAdapter<AppService> _logger; 
+        private readonly LoggerContext _loggerContext;
+
+        public MessageProcessor(ILoggerAdapter<AppService> logger, LoggerContext loggerContext){
```
Comment (tobiasbrown) [inline subscribers/logs/logger/MessageProcessor.cs]:
```suggestion
        public MessageProcessor(ILoggerAdapter<AppService> logger, LoggerContext loggerContext) { 
```

File: subscribers/logs/logger/MessageProcessor.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+    public class MessageProcessor: Exception, IMessageProcessor
+    {
+        private readonly ILoggerAdapter<AppService> _logger; 
+        private readonly LoggerContext _loggerContext;
+
+        public MessageProcessor(ILoggerAdapter<AppService> logger, LoggerContext loggerContext){ 
+            _logger = logger;
+            _loggerContext = loggerContext;
+        }
+        public void Process(Message message) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/MessageProcessor.cs:18]:
```suggestion

        public void Process(Message message) {
```

File: subscribers/slack/worker.tests/UserMessageProcessorTest.cs [hunk 21]
```diff
@@ -1,9 +1,8 @@
+using System.Threading.Tasks;
```
Comment (malmas18) [inline subscribers/slack/worker.tests/UserMessageProcessorTest.cs]:
```suggestion
```

File: subscribers/logs/logger/Program.cs [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+                    services.AddOptions();
+                    services.AddSingleton<IHostedService, AppService>();
+                    services.AddSingleton(typeof(ILoggerAdapter<>),typeof(LoggerAdapter<>));
+                    services.AddTransient<IMessageProcessor, MessageProcessor>();
+                    services.Configure<AppConfig>(ac => {
+                        ac.AwsSqsAccessKeyId = Environment.GetEnvironmentVariable("AWS_SQS_ACCESS_KEY_ID");
+                        ac.AwsSqsQueueUrl = Environment.GetEnvironmentVariable("AWS_SQS_QUEUE_URL");
+                        ac.AwsSqsServiceUrl = Environment.GetEnvironmentVariable("AWS_SQS_SERVICE_URL");
+                        var awsSqsRegion = Environment.GetEnvironmentVariable("AWS_SQS_REGION");
+                        if (string.IsNullOrWhiteSpace(awsSqsRegion) == false) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/Program.cs:55]:
```suggestion
                        if (!string.IsNullOrWhiteSpace(awsSqsRegion)) {
```

File: subscribers/logs/logger/Program.cs [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+                            ac.AwsSqsRegion = awsSqsRegion;
+                        }
+                        ac.AwsSqsSecretAccessKey = Environment.GetEnvironmentVariable("AWS_SQS_SECRET_ACCESS_KEY");
+                        ac.ConnectionString = Environment.GetEnvironmentVariable("CONNECTION_STRING");
+
+                        var workIntervalInSeconds = Environment.GetEnvironmentVariable("WORK_INTERVAL_IN_SECONDS");
+                        if (string.IsNullOrWhiteSpace(workIntervalInSeconds) == false) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/Program.cs:62]:
```suggestion
                        if (!string.IsNullOrWhiteSpace(workIntervalInSeconds)) {
```

File: subscribers/logs/logger/Program.cs [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+                            ac.WorkIntervalInSeconds = int.Parse(workIntervalInSeconds);
+                        }
+                        var awsSqsLongPollTimeInSeconds = Environment.GetEnvironmentVariable("AWS_SQS_LONG_POLL_TIME_IN_SECONDS");
+                        if (string.IsNullOrWhiteSpace(awsSqsLongPollTimeInSeconds) == false) {
```
Comment (tobiasbrown) [inline subscribers/logs/logger/Program.cs:66]:
```suggestion
                        if (!string.IsNullOrWhiteSpace(awsSqsLongPollTimeInSeconds)) {
```

**Review by malmas18 — APPROVED** [review]

File: subscribers/logs/logger.tests/LoggerTests.cs (hunk 9 above)
Comment (resh-ie) [reply subscribers/logs/logger.tests/LoggerTests.cs]: fixed to add the right c# naming convengtion

File: subscribers/slack/worker.tests/UserMessageProcessorTest.cs (hunk 21 above)
Comment (resh-ie) [reply subscribers/slack/worker.tests/UserMessageProcessorTest.cs]: done

File: subscribers/logs/logger/AppService.cs (hunk 18 above)
Comment (resh-ie) [reply subscribers/logs/logger/AppService.cs]: fixed
